### ISSUE TITLE
gh-ost seems blocked when Backlog get 100/100

### ISSUE THREAD

**Issue by sumonsun** [github/gh-ost#348]
#### when Backlog get 100/100,gh-ost didn't work,can we increase this parameter?

```
gh-ost --allow-on-master --max-load=Threads_running=30 --critical-load=Threads_running=45 --chunk-size=1000 --max-lag-millis=1500 --initially-drop-old-table --initially-drop-ghost-table --initially-drop-socket-file --ok-to-drop-table --host="10.0.xxx" --port=3306 --user="xxx" --password="xxx" --database="test" --table="comm_go_ost" --verbose --alter="add column testa varchar(50)" --switch-to-rbr --cut-over=default --default-retries=120 --panic-flag-file=/tmp/ghost.panic.flag --postpone-cut-over-flag-file=/tmp/ghost.postpone.flag  --throttle-query="select 0" --execute
	
	
Copy: 4979000/22690750 21.9%; Applied: 612609; Backlog: 100/100; Time: 11m30s(total), 11m30s(copy); streamer: mysql-bin.000187:386543989; State: migrating; ETA: 40m54s
Copy: 4979000/22690750 21.9%; Applied: 641517; Backlog: 100/100; Time: 12m0s(total), 12m0s(copy); streamer: mysql-bin.000187:390623293; State: migrating; ETA: 42m41s
Copy: 4979000/22690750 21.9%; Applied: 670192; Backlog: 100/100; Time: 12m30s(total), 12m30s(copy); streamer: mysql-bin.000187:394667040; State: migrating; ETA: 44m27s
Copy: 4979000/22690750 21.9%; Applied: 698549; Backlog: 100/100; Time: 13m0s(total), 13m0s(copy); streamer: mysql-bin.000187:398671937; State: migrating; ETA: 46m14s
Copy: 4979000/22690750 21.9%; Applied: 727394; Backlog: 100/100; Time: 13m30s(total), 13m30s(copy); streamer: mysql-bin.000187:402753149; State: migrating; ETA: 48m1s
```

**Comment by shlomi-noach** [comment]
If the backlog is `100/100` that means `gh-ost` is unable to apply the binlog events it reads from the binary logs onto the _ghost_ table.

Typically this is because `gh-ost` is throttling, however this isn't your case. In you case, there is no throttling, and `gh-ost` advances in the binary log: `mysql-bin.000187:386543989`, `mysql-bin.000187:390623293`, ...

This indicates `gh-ost` doesn't have the write capacity to apply the binlog events before new ones arrive.

Increasing the backlog buffer's size would not make a difference, since it would also get filled anyhow. Besides, `gh-ost` prioritizes applying binlog event changes over row-copy.

So the question at hand is: does your server has enough write capacity to run a migration? At the end of the day, you need the capacity to copy `22690750` rows + ongoing changes. If you don't have such capacity, there is no possibility for a migration.

The good news is, maybe you're just suffering from rush hour load. If during night time or weekend your load decreases, `gh-ost` will eventually find the capacity to resume copying and completing the migration.

**Comment by sumonsun** [comment]
thanks for your quickly reply.
You are right ,if the speed of applied events behind the speed of event produce ,the hit the pain point,I increase backlog to 10000000, then I stop write load on that server, backlog reduce slowly ,applied events speed about 1000 events/s. so I think can we increase the capacity of applied events.
Copy: 0/169923421 0.0%; Applied: 916066; Backlog: 3066373/10000000; Time: 15m30s(total), 15m30s(copy); streamer: mysql-bin.000209:478100492; State: migrating; ETA: N/A
[MySQL] 2016/12/27 17:07:42 statement.go:27: Invalid Connection
Copy: 0/169923421 0.0%; Applied: 946733; Backlog: 3035706/10000000; Time: 16m0s(total), 16m0s(copy); streamer: mysql-bin.000209:488976943; State: migrating; ETA: N/A
Copy: 0/169923421 0.0%; Applied: 977568; Backlog: 3004871/10000000; Time: 16m30s(total), 16m30s(copy); streamer: mysql-bin.000209:499912858; State: migrating; ETA: N/A

**Comment by shlomi-noach** [comment]
Again, I advise against increasing the backlog size, and this is why it's not configurable anyway. Stopping write load will work just as well with backlog of 100. 
10,000,000 also implies a lot of consumed memory. 

Moreover , from your log it doesn't seem like this made a difference. Rowcopy still remains static. 

Lastly, in both your examples you have been running for less than 20 minutes. I'd be more interested to see how that proceeds over the next hour, two, or more.

**Comment by krzysztof-ksiazek** [comment]
@shlomi-noach this seems to be in line with what I see in my setup that I reported to you via email. Here's a bit more details. I'm using m4.10xl instances with 5k IOPS io1 volumes, this should guarantee throughput of ~78MB/s with 16KB pages or 500MB/s with larger chunks.

Here's the output from gh-ost + some data on CPU and I/O load on the system.

https://gist.github.com/krzysztof-ksiazek/37149a420cbf2517afc28fbbf63e6c50

As you can see, gh-ost processes on average ~400 events per second which is not enough to keep up with three threads sysbench load like:

sysbench --test=/root/sysbench/sysbench/tests/db/oltp.lua --num-threads=3 --max-requests=0 --max-time=600 --mysql-port=6033 --mysql-host=127.0.0.1 --mysql-user=sbtest --mysql-password=sbtest --oltp-tables-count=1 --mysql-db=sbtest1 --report-interval=1 --oltp-table-size=10000000 --oltp-skip-trx=on run

[   1s] threads: 3, tps: 0.00, reads: 2050.68, writes: 575.91, response time: 28.03ms (95%), errors: 0.00, reconnects:  0.00
[   2s] threads: 3, tps: 0.00, reads: 1940.03, writes: 556.01, response time: 29.99ms (95%), errors: 0.00, reconnects:  0.00
[   3s] threads: 3, tps: 0.00, reads: 1581.00, writes: 455.00, response time: 30.86ms (95%), errors: 0.00, reconnects:  0.00
[   4s] threads: 3, tps: 0.00, reads: 1875.99, writes: 534.00, response time: 31.92ms (95%), errors: 0.00, reconnects:  0.00
[   5s] threads: 3, tps: 0.00, reads: 1710.01, writes: 493.00, response time: 32.99ms (95%), errors: 0.00, reconnects:  0.00

As a result, gh-ost won't be able to perform migration as long as load won't decrease below the 400 events per second and gh-ost will be able to clean backlog.

I understand this is a worst case scenario - bunch of very small events - PK UPDATE/DELETE or single row inserts. It's not uncommon to see ~400+ DML's per second so I'm really curious if this is something wrong with my environment or just that's how gh-ost performs at this moment.

**Comment by shlomi-noach** [comment]
@krzysztof-ksiazek thank you for the metrics, that's helpful. 

I don't see that I have a production case of continuous `400` events/sec on a given table, and so I will need to benchmark this artificially.

If `gh-ost` is slow to apply binlog events, this could be either:

1. slow to fetch binlog events
2. slow to sift through events
3. slow to process a DML event
4. slow to apply DML event onto _ghost_ table 

If your case is similar to the original @sumonsun case, then it isn't (1): when @sumonsun increased buffer/backlog size, `gh-ost` clearly processed such events from the binary log.
This doesn't seem like (2) or (3) because your output suggests reasonable CPU usage.

So, (4): I'd be interested to get some metrics on the _ghost_ table: waits on table, locks?

If (4) is indeed the case, then a potential fix would be to apply DML in batched commits. This was suggested by the Oracle engineers in our meeting but at the time I saw no need for that.

If further evidence supports (4) then I'll pursue that; so, basically, this is a request for more metrics to blame (1), (2), (3) or (4).

**Comment by krzysztof-ksiazek** [comment]
I'm not entirely sure what exact data you may need so if you'd like to see something more, let me know.
Below are some stats from PS and data from couple of subsequent pt-stalk executions.

https://gist.github.com/krzysztof-ksiazek/8abcd43a600d26b077eabab79c8e6e5a
https://drive.google.com/file/d/0B3on9veKOlkDVWprSmlpWlpxMWc/view?usp=sharing

**Comment by shlomi-noach** [comment]
@sumonsun, @krzysztof-ksiazek: will you kindly try this binary: 
[gh-ost-binary-linux-20170104075031.tar.gz](https://github.com/github/gh-ost/files/683985/gh-ost-binary-linux-20170104075031.tar.gz), which is a build of https://github.com/github/gh-ost/issues/352 ?
This introduces `--dml-batch-size`, default `10`, and assuming the original problem is commit capacity, this should help. Valid range for `--dml-batch-size` is `1..100`.

Please let me know how goes.

**Comment by shlomi-noach** [comment]
Please try this binary: 
[gh-ost-binary-linux-20170105104155.tar.gz](https://github.com/github/gh-ost/files/686785/gh-ost-binary-linux-20170105104155.tar.gz)

**Comment by krzysztof-ksiazek** [comment]
I did some quick tests yesterday with the previous build, it looked much better. I'll run some more proper testing with the latest binary and share it here.

**Comment by shlomi-noach** [comment]
Outstanding. I'm meanwhile testing in production ; I will surely run these tests for a few days.

**Comment by krzysztof-ksiazek** [comment]
https://gist.github.com/krzysztof-ksiazek/b9bcedab0c08963513d7a810a7755c83

As I said, it looks much better - bottleneck is now in replication performance, not in how fast gh-ost can apply binlog events. Default dml-batch-size seems to work just fine, although increasing it to 100 allows for slightly more events to be processed at once, even up to ~1.5-2k.

**Comment by shlomi-noach** [comment]
I'm going to close this based on response by @krzysztof-ksiazek. @sumonsun, when you happen to test this, please drop a note, and of course please reopen is you still see problems.

This is fixed by #352 and released in https://github.com/github/gh-ost/releases/tag/v1.0.34

**Comment by sumonsun** [comment]
Thanks Shlomi

I did benchmark by this new verison, work fine,performance is perfect.apply speed was very fast .

thanks so much ,I will do more benchmark to valid this new version.
